File: src/day07.rs
pub fn part1(input: String) -> i64 {
    return search(input, |a, b| (a - b).abs());
}

pub fn part2(input: String) -> i64 {
    return search(input, |a, b| (a - b).abs() * ((a - b).abs() + 1) / 2);
}

pub fn search(input: String, cost: fn(a: i32, b: i32) -> i32) -> i64 {
    let values: Vec<i32> = input
        .split(',')
        .map(|s| s.trim().parse::<i32>().unwrap())
        .collect();
    let min_value = values.iter().min().unwrap();
    let max_value = values.iter().max().unwrap();
    let h_sum =
        (*min_value..*max_value).map(|h| (h, values.iter().map(|v| cost(*v, h)).sum::<i32>()));
    let (_, sum) = h_sum.min_by(|(_, s1), (_, s2)| s1.cmp(s2)).unwrap();
    return sum as i64;
}

File: go/go07/day07.go
package main

import (
	"fmt"
	"io/ioutil"
	"math"
	"strconv"
	"strings"
	"time"
)

type fn func(int, int) int

func search(values []int, cost fn) int {
	minValue, maxValue := math.MaxInt, 0
	for _, value := range values {
		if value < minValue {
			minValue = value
		}
		if value > maxValue {
			maxValue = value
		}
	}

	minSum := math.MaxInt

	for h := minValue; h <= maxValue; h++ {
		sum := 0
		for _, value := range values {
			sum += cost(h, value)
		}
		if sum < minSum {
			minSum = sum
		}
	}
	return minSum
}

func Part1(input string) int {
	values := make([]int, 0)
	for _, in := range strings.Split(strings.TrimSuffix(input, "\n"), ",") {
		v, err := strconv.Atoi(in)
		if err != nil {
			panic(err)
		}
		values = append(values, v)
	}
	return search(values, func(a, b int) int {
		return abs(a - b)
	})
}

func Part2(input string) int {
	values := make([]int, 0)
	for _, in := range strings.Split(strings.TrimSuffix(input, "\n"), ",") {
		v, _ := strconv.Atoi(in)
		values = append(values, v)
	}
	return search(values, func(a, b int) int {
		n := abs(a - b)
		return n * (n + 1) / 2
	})
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	content, _ := ioutil.ReadFile("../../inputs/day07.txt")

	start := time.Now()
	fmt.Println("part1: ", Part1(string(content)))
	fmt.Println(time.Since(start))

	start = time.Now()
	fmt.Println("part2: ", Part2(string(content)))
	fmt.Println(time.Since(start))
}
